### DISCUSSION TITLE
How do you design API around request context?

### DISCUSSION THREAD

**Discussion by radoslavkarlik** [oscartbeaumont/rspc#75] [General]
Hi, my app is local only so I store all data in my own services. When returning them, I return references to them. I think it should be possible to use the references, serialize their values and return the references back. I am new to Rust and perhaps my question is rather a design problem but I feel lost now.

Compilation error
```
cannot return value referencing function parameter `ctx`
returns a value referencing data owned by the current function
```

```rust
//  also stored as `Arc<Context>` and cloned for each request
struct Context {
    accounts: Arc<Mutex<Accounts>>,
}
impl Accounts {
    fn all(&self) -> Vec<&Account> {...} // should it not return references?
}

.query("accounts", |t| {
    t(|ctx, _: ()| {
        let accounts = ctx.accounts.lock().unwrap().all();
        Ok(accounts)
    })
})
```

Are we expected to not use any references and always move data with ownership?
From the example: https://rspc.dev/server/router#attaching-procedures
```rust
User::get_all() // returns references or with ownership?
```

**Comment by radoslavkarlik** [comment]
I ended up cloning stuff in the closures as that is probably how Rust APIs should work as the values behind references could change between finishing the closure code and accessing them to serialize.

**Comment by oscartbeaumont** [comment]
Using `Arc` and cloning is currently the best way. Due to an `Arc` only cloning the reference and not the data it is a relatively cheap operation so `Arc` cloning is the recommended way to get data into the request context. I would like to allow using references but I don't know that it is going to be possible to express the lifetimes.

Whatever you return from your rspc resolver will most likely need to be owned to satisfy the borrow checker so I would just have `Account::all` clone the array so the values are owned and can be easily returned.
